src/components/submission/diff-item.vue

```diff
@@ -0,0 +1,267 @@
+      if (this.entry.old === null && typeof (this.entry.new) === 'object')
+        return false;
+      if (this.entry.new === null && typeof (this.entry.old) === 'object')
+        return false;
+      return true;
+    },
+    visiblePath() {
+      // If the change is atomic, it will split the path into the field name
+      // of the changed leaf node and the preceeding part of the path (which
+      // could be empty for a short path).
+      // Otherwise, it will show the whole path of the changed subtree.
+      if (!this.isAtomicChange)
+        return this.entry.path;
+      return this.entry.path.slice(0, this.entry.path.length - 1);
+    },
+    fieldName() {
+      return last(this.entry.path);
+    },
+    nestedClass() {
+      return this.parentPath.length === 0 ? 'outer-item' : 'inner-item';
+    },
+    typeOfChange() {
+      // Check which value is null (old or new) to determine the
+      // type of change or edit.
+      // This is only used for non-atomic (nested) changes so
+      // it will always be one or the other.
+      return (this.entry.old === null ? 'added' : 'deleted');
+    },
+    nestedDiffs() {
+      return this.flattenDiff(this.entry.new || this.entry.old, this.typeOfChange);
+    },
+    isBinary() {
+      // Compares path as array by converting it to /field1/field2 string and checks
+      // if it is for a binary field
+      const fullPath = this.parentPath.concat(this.entry.path);
+      const fullPathStr = fullPath.map((field) => (Array.isArray(field) ? field[0] : field)).join('/');
+      const basicPath = `/${fullPathStr}`;
+      if (this.binaryFieldPaths.has(basicPath))
+        return true;
+      return false;
+    }
+  },
+  methods: {
+    flattenDiff(node, typeOfChange, prefix = []) {
+      // Flattens the inner diff (JSON of an entire added/removed tree structure)
+      // by recursively traversing the tree
+      const changes = [];
+      // eslint-disable-next-line guard-for-in
+      for (const k in node) {
+        const path = prefix.concat(k);
+        if (Array.isArray(node)) {
+          path.pop(); // pop off the index (e.g. 1 of toy #1)
+          path.pop(); // pop off the name (e.g. toy of toy #1)
+          path.push([prefix[prefix.length - 1], Number.parseInt(k, 10)]);
+        }
+        if (typeof (node[k]) === 'object') {
+          changes.push(...this.flattenDiff(node[k], typeOfChange, path));
+        } else if (typeOfChange === 'added') {
+          changes.push({ path, old: null, new: node[k] });
+        } else {
+          changes.push({ path, old: node[k], new: null });
+        }
+      }
+      return changes;
+    },
+    binaryHref(value) {
+      return apiPaths.submissionAttachment(
+        this.projectId,
+        this.xmlFormId,
+        false,
+        this.instanceId,
+        value
+      );
+    },
+    formatData(content) {
+      const length = 240;
+      return content.length > length ? content.slice(0, length).concat('...') : content;
+    }
```

Comment:
Rather than doing this in the JavaScript, I think it'd be better to do it in the CSS using the `text-overflow-ellipsis` mixin. Here are a couple of options that might work:

Option 1: use `inline-block`

```scss
.data-old, .data-new {
  @include text-overflow-ellipsis;
  display: inline-block;
  max-width: 250px; // or a different number
}
```

Option 2: use flexbox

```scss
// You might need some additional flexbox options?
.old-to-new { display: flex; }

.data-old, .data-new {
  @include text-overflow-ellipsis;
  max-width: 250px; // or a different number
}
```

Reply:
I'm trying out the flexbox route and it's ok except I wish I could get it to expand to fill the whole space while still doing the overflow ellipses trick.

<img width="1192" alt="Screen Shot 2021-07-27 at 4 24 02 PM" src="https://user-images.githubusercontent.com/76205/127240024-15fd91c9-66c4-4129-a38e-020417af10f9.png">

It also looks pretty weird when the page is too small and needs to scroll horizontally.

<img width="652" alt="Screen Shot 2021-07-27 at 4 25 50 PM" src="https://user-images.githubusercontent.com/76205/127240192-1ce03700-bdb5-4bf9-987f-c8ecfd37e23e.png">

Reply:
> I'm trying out the flexbox route and it's ok except I wish I could get it to expand to fill the whole space while still doing the overflow ellipses trick.

I'm not totally sure, but I think that's possible using flexbox! Maybe remove the `width` property from `.data-old` and `.data-new` and instead specify the `flex` property (or maybe just `flex-grow`)? I think it should be possible to keep the icon a fixed width, then divvy up the remaining space to `.data-old` and `.data-new`.

This also feels like a nice-to-have, so if it ends up being challenging, I think we could come back to it.

> It also looks pretty weird when the page is too small and needs to scroll horizontally.

I think this comes from specifying a fixed `width`. If you're able to solve the above and get `.data-old` and `.data-new` to divvy up the full width, I think that'll also solve this issue. Alternatively, you could try replacing `width` with `max-width`.

That said, Frontend also isn't designed for smaller viewports, and there'll be other overflow issues if the viewport is too narrow. We generally assume that the viewport is at least 1000px wide. As long as things look OK at 1000px, I think that's probably fine!

Reply:
Going with the condensed 250px wide & ellipsis overflow solution for now since getting something that consistently worked and looked ok was challenging.